File: CS/MVVMExpenses/Common/DataModel/EntityFramework/DbRepository.cs
using System;
using System.Linq;
using System.Data;
using System.Data.Entity;
using System.Linq.Expressions;
using System.Collections.Generic;
using MVVMExpenses.Common.Utils;
using MVVMExpenses.Common.DataModel;
using System.Data.Entity.Validation;
using System.Data.Entity.Infrastructure;

namespace MVVMExpenses.Common.DataModel.EntityFramework {
    /// <summary>
    /// A DbRepository is a IRepository interface implementation representing the collection of all entities in the unit of work, or that can be queried from the database, of a given type. 
    /// DbRepository objects are created from a DbUnitOfWork using the GetRepository method. 
    /// DbRepository provides only write operations against entities of a given type in addition to the read-only operation provided DbReadOnlyRepository base class.
    /// </summary>
    /// <typeparam name="TEntity">Repository entity type.</typeparam>
    /// <typeparam name="TPrimaryKey">Entity primary key type.</typeparam>
    /// <typeparam name="TDbContext">DbContext type.</typeparam>
    public class DbRepository<TEntity, TPrimaryKey, TDbContext> : DbReadOnlyRepository<TEntity, TDbContext>, IRepository<TEntity, TPrimaryKey>
        where TEntity : class
        where TDbContext : DbContext {

        readonly Expression<Func<TEntity, TPrimaryKey>> getPrimaryKeyExpression;
        readonly EntityTraits<TEntity, TPrimaryKey> entityTraits;

        /// <summary>
        /// Initializes a new instance of DbRepository class.
        /// </summary>
        /// <param name="unitOfWork">Owner unit of work that provides context for repository entities.</param>
        /// <param name="dbSetAccessor">Function that returns DbSet entities from Entity Framework DbContext.</param>
        /// <param name="getPrimaryKeyExpression">Lambda-expression that returns entity primary key.</param>
        public DbRepository(DbUnitOfWork<TDbContext> unitOfWork, Func<TDbContext, DbSet<TEntity>> dbSetAccessor, Expression<Func<TEntity, TPrimaryKey>> getPrimaryKeyExpression)
            : base(unitOfWork, dbSetAccessor) {
            this.getPrimaryKeyExpression = getPrimaryKeyExpression;
            this.entityTraits = ExpressionHelper.GetEntityTraits(this, getPrimaryKeyExpression);
        }

        protected virtual TEntity CreateCore(bool add = true) {
            TEntity newEntity = DbSet.Create();
            if(add) {
                DbSet.Add(newEntity);
            }
            return newEntity;
        }

        protected virtual void UpdateCore(TEntity entity) {
        }

        protected virtual EntityState GetStateCore(TEntity entity) {
            return GetEntityState(Context.Entry(entity).State);
        }

        static EntityState GetEntityState(System.Data.Entity.EntityState entityStates) {
            switch(entityStates) {
                case System.Data.Entity.EntityState.Added:
                    return EntityState.Added;
                case System.Data.Entity.EntityState.Deleted:
                    return EntityState.Deleted;
                case System.Data.Entity.EntityState.Detached:
                    return EntityState.Detached;
                case System.Data.Entity.EntityState.Modified:
                    return EntityState.Modified;
                case System.Data.Entity.EntityState.Unchanged:
                    return EntityState.Unchanged;
                default:
                    throw new NotImplementedException();
            }
        }


        protected virtual TEntity FindCore(TPrimaryKey primaryKey) {
            return DbSet.Find(primaryKey);
        }

        protected virtual void RemoveCore(TEntity entity) {
            try {
                DbSet.Remove(entity);
            }
            catch(DbEntityValidationException ex) {
                throw DbExceptionsConverter.Convert(ex);
            }
            catch(DbUpdateException ex) {
                throw DbExceptionsConverter.Convert(ex);
            }
        }

        protected virtual TEntity ReloadCore(TEntity entity) {
            Context.Entry(entity).Reload();
            return FindCore(GetPrimaryKeyCore(entity));
        }
        protected virtual TPrimaryKey GetPrimaryKeyCore(TEntity entity) {
            return entityTraits.GetPrimaryKey(entity);
        }

        protected virtual void SetPrimaryKeyCore(TEntity entity, TPrimaryKey primaryKey) {
            var setPrimaryKeyAction = entityTraits.SetPrimaryKey;
            setPrimaryKeyAction(entity, primaryKey);
        }

        #region IRepository
        TEntity IRepository<TEntity, TPrimaryKey>.Find(TPrimaryKey primaryKey) {
            return FindCore(primaryKey);
        }

        void IRepository<TEntity, TPrimaryKey>.Add(TEntity entity) {
            DbSet.Add(entity);
        }

        void IRepository<TEntity, TPrimaryKey>.Remove(TEntity entity) {
            RemoveCore(entity);
        }

        TEntity IRepository<TEntity, TPrimaryKey>.Create(bool add) {
            return CreateCore(add);
        }

        void IRepository<TEntity, TPrimaryKey>.Update(TEntity entity) {
            UpdateCore(entity);
        }

        EntityState IRepository<TEntity, TPrimaryKey>.GetState(TEntity entity) {
            return GetStateCore(entity);
        }

        TEntity IRepository<TEntity, TPrimaryKey>.Reload(TEntity entity) {
            return ReloadCore(entity);
        }

        Expression<Func<TEntity, TPrimaryKey>> IRepository<TEntity, TPrimaryKey>.GetPrimaryKeyExpression {
            get { return this.getPrimaryKeyExpression; }
        }

        void IRepository<TEntity, TPrimaryKey>.SetPrimaryKey(TEntity entity, TPrimaryKey primaryKey) {
            SetPrimaryKeyCore(entity, primaryKey);
        }

        TPrimaryKey IRepository<TEntity, TPrimaryKey>.GetPrimaryKey(TEntity entity) {
            return GetPrimaryKeyCore(entity);
        }

        bool IRepository<TEntity, TPrimaryKey>.HasPrimaryKey(TEntity entity) {
            return entityTraits.HasPrimaryKey(entity);
        }
        #endregion
    }
}

File: VB/MVVMExpenses/Common/DataModel/EntityFramework/DbRepository.vb
Imports System
Imports System.Linq
Imports System.Data
Imports System.Data.Entity
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports Common.Utils
Imports Common.DataModel
Imports System.Data.Entity.Validation
Imports System.Data.Entity.Infrastructure
Namespace Common.DataModel.EntityFramework
    ''' <summary>
    ''' A DbRepository is a IRepository interface implementation representing the collection of all entities in the unit of work, or that can be queried from the database, of a given type. 
    ''' DbRepository objects are created from a DbUnitOfWork using the GetRepository method. 
    ''' DbRepository provides only write operations against entities of a given type in addition to the read-only operation provided DbReadOnlyRepository base class.
    ''' </summary>
    ''' <typeparam name="TEntity">Repository entity type.</typeparam>
    ''' <typeparam name="TPrimaryKey">Entity primary key type.</typeparam>
    ''' <typeparam name="TDbContext">DbContext type.</typeparam>
    Public Class DbRepository(Of TEntity As Class, TPrimaryKey, TDbContext As DbContext)
        Inherits DbReadOnlyRepository(Of TEntity, TDbContext)
        Implements IRepository(Of TEntity, TPrimaryKey)
        Private ReadOnly _getPrimaryKeyExpression As Expression(Of Func(Of TEntity, TPrimaryKey))
        Private ReadOnly _entityTraits As EntityTraits(Of TEntity, TPrimaryKey)
        ''' <summary>
        ''' Initializes a new instance of DbRepository class.
        ''' </summary>
        ''' <param name="unitOfWork">Owner unit of work that provides context for repository entities.</param>
        ''' <param name="dbSetAccessor">Function that returns DbSet entities from Entity Framework DbContext.</param>
        ''' <param name="getPrimaryKeyExpression">Lambda-expression that returns entity primary key.</param>
        Public Sub New(ByVal unitOfWork As DbUnitOfWork(Of TDbContext), ByVal dbSetAccessor As Func(Of TDbContext, DbSet(Of TEntity)), ByVal getPrimaryKeyExpression As Expression(Of Func(Of TEntity, TPrimaryKey)))
            MyBase.New(unitOfWork, dbSetAccessor)
            Me._getPrimaryKeyExpression = getPrimaryKeyExpression
            Me._entityTraits = ExpressionHelper.GetEntityTraits(Me, getPrimaryKeyExpression)
        End Sub
        Protected Overridable Function CreateCore(Optional ByVal add As Boolean = True) As TEntity
            Dim newEntity As TEntity = DbSet.Create()
            If add Then
                DbSet.Add(newEntity)
            End If
            Return newEntity
        End Function
        Protected Overridable Sub UpdateCore(ByVal entity As TEntity)
        End Sub
        Protected Overridable Function GetStateCore(ByVal entity As TEntity) As EntityState
            Return GetEntityState(Context.Entry(entity).State)
        End Function
        Private Shared Function GetEntityState(ByVal entityStates As System.Data.Entity.EntityState) As EntityState
            Select Case entityStates
                Case System.Data.Entity.EntityState.Added
                    Return EntityState.Added
                Case System.Data.Entity.EntityState.Deleted
                    Return EntityState.Deleted
                Case System.Data.Entity.EntityState.Detached
                    Return EntityState.Detached
                Case System.Data.Entity.EntityState.Modified
                    Return EntityState.Modified
                Case System.Data.Entity.EntityState.Unchanged
                    Return EntityState.Unchanged
                Case Else
                    Throw New NotImplementedException()
            End Select
        End Function
        Protected Overridable Function FindCore(ByVal primaryKey As TPrimaryKey) As TEntity
            Return DbSet.Find(primaryKey)
        End Function
        Protected Overridable Sub RemoveCore(ByVal entity As TEntity)
            Try
                DbSet.Remove(entity)
            Catch ex As DbEntityValidationException
                Throw DbExceptionsConverter.Convert(ex)
            Catch ex As DbUpdateException
                Throw DbExceptionsConverter.Convert(ex)
            End Try
        End Sub
        Protected Overridable Function ReloadCore(ByVal entity As TEntity) As TEntity
            Context.Entry(entity).Reload()
            Return FindCore(GetPrimaryKeyCore(entity))
        End Function
        Protected Overridable Function GetPrimaryKeyCore(ByVal entity As TEntity) As TPrimaryKey
            Return _entityTraits.GetPrimaryKey(entity)
        End Function
        Protected Overridable Sub SetPrimaryKeyCore(ByVal entity As TEntity, ByVal primaryKey As TPrimaryKey)
            Dim setPrimaryKeyAction = _entityTraits.SetPrimaryKey
            setPrimaryKeyAction(entity, primaryKey)
        End Sub
        Private Function Find(ByVal primaryKey As TPrimaryKey) As TEntity Implements IRepository(Of TEntity, TPrimaryKey).Find
            Return FindCore(primaryKey)
        End Function
        Private Sub Add(ByVal entity As TEntity) Implements IRepository(Of TEntity, TPrimaryKey).Add
            DbSet.Add(entity)
        End Sub
        Private Sub Remove(ByVal entity As TEntity) Implements IRepository(Of TEntity, TPrimaryKey).Remove
            RemoveCore(entity)
        End Sub
        Private Function Create(Optional ByVal add As Boolean = True) As TEntity Implements IRepository(Of TEntity, TPrimaryKey).Create
            Return CreateCore(add)
        End Function
        Private Sub Update(ByVal entity As TEntity) Implements IRepository(Of TEntity, TPrimaryKey).Update
            UpdateCore(entity)
        End Sub
        Private Function GetState(ByVal entity As TEntity) As EntityState Implements IRepository(Of TEntity, TPrimaryKey).GetState
            Return GetStateCore(entity)
        End Function
        Private Function Reload(ByVal entity As TEntity) As TEntity Implements IRepository(Of TEntity, TPrimaryKey).Reload
            Return ReloadCore(entity)
        End Function
        Private ReadOnly Property GetPrimaryKeyExpression As Expression(Of Func(Of TEntity, TPrimaryKey)) Implements IRepository(Of TEntity, TPrimaryKey).GetPrimaryKeyExpression
            Get
                Return Me._getPrimaryKeyExpression
            End Get
        End Property
        Private Sub SetPrimaryKey(ByVal entity As TEntity, ByVal primaryKey As TPrimaryKey) Implements IRepository(Of TEntity, TPrimaryKey).SetPrimaryKey
            SetPrimaryKeyCore(entity, primaryKey)
        End Sub
        Private Function GetPrimaryKey(ByVal entity As TEntity) As TPrimaryKey Implements IRepository(Of TEntity, TPrimaryKey).GetPrimaryKey
            Return GetPrimaryKeyCore(entity)
        End Function
        Private Function HasPrimaryKey(ByVal entity As TEntity) As Boolean Implements IRepository(Of TEntity, TPrimaryKey).HasPrimaryKey
            Return _entityTraits.HasPrimaryKey(entity)
        End Function
    End Class
End Namespace
